test/unit/api.test.js

```diff
@@ -0,0 +1,203 @@
+  ['Admin', () => new mongodbLegacy.Admin(db)],
+  ['FindCursor', () => new mongodbLegacy.FindCursor(client, namespace)],
+  ['ListCollectionsCursor', () => new mongodbLegacy.ListCollectionsCursor(db, {})],
+  ['ListIndexesCursor', () => new mongodbLegacy.ListIndexesCursor(collection)],
+  ['AggregationCursor', () => new mongodbLegacy.AggregationCursor(client, namespace)],
+  ['ChangeStream', () => new mongodbLegacy.ChangeStream(client)],
+  ['GridFSBucket', () => new mongodbLegacy.GridFSBucket(db)],
+  ['Collection', () => new mongodbLegacy.Collection(db, 'pets')],
+  ['Db', () => new mongodbLegacy.Db(client, 'animals')],
+  ['MongoClient', () => new mongodbLegacy.MongoClient(iLoveJs)]
+]);
+
+function* generateTests() {
+  for (const [className, getInstance] of OVERRIDDEN_CLASSES_GETTER) {
+    // For each of the overridden classes listed above
+    /** @type {{ className: string; method: string; returnType: string; special?: string; }[]} */
+    let methods = classesToMethods.get(className);
+    if (methods == null) methods = [];
+
+    for (const { method, special, possibleCallbackPositions } of methods) {
+      // for each of the methods on the overridden class
+      if (typeof special === 'string' && special.length !== 0) {
+        // If there is a special handling reason this method should be manually tested elsewhere
+        continue;
+      }
+
+      // Make the readable API name and construct an instance for testing
+      const apiName = `${className}.${method}`;
+      const instance = getInstance();
+      yield {
+        className,
+        method,
+        instance,
+        apiName,
+        possibleCallbackPositions,
+        makeStub: superPromise => {
+          return sinon.stub(mongodbDriver[className].prototype, method).returns(superPromise);
+        }
+      };
+    }
+  }
+}
+
+describe('Maybe Callback', () => {
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  for (const {
+    apiName,
+    instance,
+    method,
+    possibleCallbackPositions,
+    makeStub
+  } of generateTests()) {
+    const functionLength = instance[method].length;
+
+    describe(`${apiName}()`, () => {
+      // Use the callback positions manually defined, or use a default of [1] / [1,2] depending on function length
+      const callbackPositions =
+        possibleCallbackPositions != null
+          ? possibleCallbackPositions
+          : functionLength < 2
+          ? [1]
+          : [1, 2];
+
+      for (const callbackPosition of callbackPositions) {
+        // Make an array that is the same length as the function under test
+        const args = Array.from({ length: functionLength }, (_, i) => i);
+        // Place the callback at the position we want to see it in
+        args[functionLength - callbackPosition] = 'callback';
+        // truncate the array
+        args.length = functionLength - (callbackPosition - 1);
+
+        it(`should support calling ${apiName}(${args.join(', ')}(undefined, result))`, async () => {
```

Comment:
can you update this and all the other test case names so that it's clear what the actual expectation is? 

the list of expectations below is very verbose, and it's hard to tell whether or not it covers whatever the intent is since that intent isn't clearly stated

you could try for something along the lines of adding an extra `context('when the parent API succeeds/fails')` and `context('when called with <whatever arguments & callback combo>')` (using whatever language you want), possibly doing a beforeach for the stub in the setup, and then having the `it` blocks actually describe the expectations e.g., `should return void`, `should call the parent API with the passed in options`, `should call the callback with (undefined, result)`

I'm open to other ideas, but we need to clearly indicate what the purpose of the test is, since "supports calling" generally just means "doesn't throw when called"

Reply:
I've nested the test names hopefully more clearly now (just in api.test.js as of writing). breaking up the assertions into more modular it blocks is difficult because setting up only one stub inside this loop can't really be done with beforeEach hooks. let me know if the current test org / titles makes more sense